tenders/common/mft.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2015-2019 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of Solo5, a sandboxed execution environment.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * mft.c: Application manifest validation and manipulation.
+ */
+
+#ifndef __SOLO5_BINDINGS__
+#include <assert.h>
+#include <string.h>
+#endif
+
+#include "mft.h"
+
+int mft_validate(struct mft *mft, size_t mft_size)
+{
+    /*
+     * Note that this function is called early in the loading process by the
+     * tenders, and must treat the manifest at (*mft) and supplied (mft_size)
+     * as untrusted data.
+     */
+    if (mft->version != MFT_VERSION)
+        return -1;
+    if (mft->entries > MFT_MAX_ENTRIES)
+        return -1;
```

Comment:
These checks dereference `mft->version` and `mft->entries` before validating that `mft_size >= sizeof(struct mft)`.
I guess that could lead to a crash under unlucky circumstances.